smithy-jmespath/src/main/java/software/amazon/smithy/jmespath/TypeChecker.java

```diff
@@ -0,0 +1,482 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import software.amazon.smithy.jmespath.ast.AndExpression;
+import software.amazon.smithy.jmespath.ast.ComparatorType;
+import software.amazon.smithy.jmespath.ast.ComparisonExpression;
+import software.amazon.smithy.jmespath.ast.CurrentExpression;
+import software.amazon.smithy.jmespath.ast.ExpressionReferenceExpression;
+import software.amazon.smithy.jmespath.ast.FieldExpression;
+import software.amazon.smithy.jmespath.ast.FilterProjectionExpression;
+import software.amazon.smithy.jmespath.ast.FlattenExpression;
+import software.amazon.smithy.jmespath.ast.FunctionExpression;
+import software.amazon.smithy.jmespath.ast.IndexExpression;
+import software.amazon.smithy.jmespath.ast.LiteralExpression;
+import software.amazon.smithy.jmespath.ast.MultiSelectHashExpression;
+import software.amazon.smithy.jmespath.ast.MultiSelectListExpression;
+import software.amazon.smithy.jmespath.ast.NotExpression;
+import software.amazon.smithy.jmespath.ast.ObjectProjectionExpression;
+import software.amazon.smithy.jmespath.ast.OrExpression;
+import software.amazon.smithy.jmespath.ast.ProjectionExpression;
+import software.amazon.smithy.jmespath.ast.SliceExpression;
+import software.amazon.smithy.jmespath.ast.Subexpression;
+
+final class TypeChecker implements ExpressionVisitor<LiteralExpression> {
+
+    private static final Map<String, FunctionDefinition> FUNCTIONS = new HashMap<>();
+
+    static {
+        FunctionDefinition.ArgValidator isAny = isType(RuntimeType.ANY);
+        FunctionDefinition.ArgValidator isString = isType(RuntimeType.STRING);
+        FunctionDefinition.ArgValidator isNumber = isType(RuntimeType.NUMBER);
+        FunctionDefinition.ArgValidator isArray = isType(RuntimeType.ARRAY);
+
+        FUNCTIONS.put("abs", new FunctionDefinition(NUMBER, isNumber));
+        FUNCTIONS.put("avg", new FunctionDefinition(NUMBER, listOfType(RuntimeType.NUMBER)));
+        FUNCTIONS.put("contains", new FunctionDefinition(
+                BOOLEAN, oneOf(RuntimeType.ARRAY, RuntimeType.STRING), isAny));
+        FUNCTIONS.put("ceil", new FunctionDefinition(NUMBER, isNumber));
+        FUNCTIONS.put("ends_with", new FunctionDefinition(NUMBER, isString, isString));
+        FUNCTIONS.put("floor", new FunctionDefinition(NUMBER, isNumber));
+        FUNCTIONS.put("join", new FunctionDefinition(STRING, isString, listOfType(RuntimeType.STRING)));
+        FUNCTIONS.put("keys", new FunctionDefinition(ARRAY, isType(RuntimeType.OBJECT)));
+        FUNCTIONS.put("length", new FunctionDefinition(
+                NUMBER, oneOf(RuntimeType.STRING, RuntimeType.ARRAY, RuntimeType.OBJECT)));
+        // TODO: Support expression reference return type validation?
+        FUNCTIONS.put("map", new FunctionDefinition(ARRAY, isType(RuntimeType.EXPRESSION_REFERENCE), isArray));
+        // TODO: support array<X|Y>
```

Comment:
This seems like it can be resolved with the tools available in the PR and is relevant for multiple functions in the list.

Reply:
Eh, I'm comfortable leaving this as a todo for now. It's probably several hours of work and testing to implement, and it's not all that important.